profile/get.go

```diff
@@ -0,0 +1,202 @@
+
+			// If we cannot find a profile, check if the request is for a special
+			// profile we can create.
+			if errors.Is(err, database.ErrNotFound) {
+				switch id {
+				case UnidentifiedProfileID:
+					profile = New(SourceLocal, UnidentifiedProfileID)
+					newProfile = true
+					err = nil
+				case SystemProfileID:
+					profile = New(SourceLocal, SystemProfileID)
+					newProfile = true
+					err = nil
+				}
+			}
+
+		case linkedPath != "":
+			// Search for profile via a linked path.
+			// Check if there already is an active and not outdated profile for
+			// the linked path.
+			profile = findActiveProfile(linkedPath)
+			if profile != nil {
+				if profile.outdated.IsSet() {
+					previousVersion = profile
+				} else {
+					return profile, nil
+				}
+			}
+			// Get from database.
+			profile, newProfile, err = findProfile(linkedPath)
+
+		default:
+			return nil, errors.New("cannot fetch profile without ID or path")
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		// Process profiles coming directly from the database.
+		// As we don't use any caching, these will be new objects.
+
+		// Mark the profile as being saved internally in order to bypass checks.
+		profile.internalSave = true
+
+		// Add a layeredProfile to local profiles.
+		if profile.Source == SourceLocal {
+			// If we are refetching, assign the layered profile from the previous version.
+			if previousVersion != nil {
+				profile.layeredProfile = previousVersion.layeredProfile
+			}
+
+			// Local profiles must have a layered profile, create a new one if it
+			// does not yet exist.
+			if profile.layeredProfile == nil {
+				profile.layeredProfile = NewLayeredProfile(profile)
+			}
+		}
+
+		// Add the profile to the currently active profiles.
+		addActiveProfile(profile)
+
+		return profile, nil
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	if p == nil {
+		return nil, false, errors.New("profile getter returned nil")
+	}
+
+	return p.(*Profile), newProfile, nil
+}
+
+// getProfile fetches the profile for the given scoped ID.
+func getProfile(scopedID string) (profile *Profile, err error) {
+	// Get profile from the database.
+	r, err := profileDB.Get(profilesDBPath + scopedID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse and prepare the profile, return the result.
+	return prepProfile(r)
+}
+
+// findProfile searches for a profile with the given linked path. If it cannot
+// find one, it will create a new profile for the given linked path.
+func findProfile(linkedPath string) (profile *Profile, new bool, err error) {
+	// Search the database for a matching profile.
+	it, err := profileDB.Query(
+		query.New(makeProfileKey(SourceLocal, "")).Where(
+			query.Where("LinkedPath", query.SameAs, linkedPath),
+		),
+	)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Only wait for the first result, or until the query ends.
+	r := <-it.Next
+	// Then cancel the query, should it still be running.
+	it.Cancel()
```

Comment:
Maybe better defer that so that it will always be canceled even in case of a panic?

Reply:
It would only move one line up and I think it's better to kill asap, as the function also does validation, which takes longer. We can save database resources by cancelling asap. If `<-it.Next` panics, `it.Cancel()` might just as well.